frontend/pipeline.yml

```
@@ -0,0 +1,302 @@
+    - changes:
+        paths: ['frontend/**/*']
+        compare_to: '${RULES_CHANGES_COMPARE_TO_REF}'
+  needs: []
+  cache:
+    key:
+      files:
+        - package-lock.json
+    paths:
+      - frontend/.npm/
+  before_script:
+    - cd frontend
+    - apk add --no-cache nodejs npm
+    - npm ci
+  script:
+    - deno run --allow-env --allow-read --allow-sys tests/licenses/licenseCheck.ts
+  tags:
+    - hetzner-amd-beefy
+
+test:frontend:unit:
+  stage: pre-test
+  image: ${CI_DEPENDENCY_PROXY_DIRECT_GROUP_IMAGE_PREFIX}/${NODE_IMAGE}
+  rules:
+    - if: $CI_COMMIT_REF_NAME !~ /^(staging|saas-[a-zA-Z0-9.]+)$/
+    - changes:
+        paths: ['frontend/**/*']
+        compare_to: '${RULES_CHANGES_COMPARE_TO_REF}'
+  needs: []
+  cache:
+    key:
+      files:
+        - package-lock.json
+    paths:
+      - frontend/.npm/
+  script:
+    - cd frontend
+    - npm ci --cache .npm --prefer-offline
+    - npm run test -- --ci --maxWorkers=2 --reporters=default --reporters=jest-junit
+  after_script:
+    - mkdir logs && cp .npm/_logs/* logs/
+  artifacts:
+    expire_in: 2w
+    paths:
+      - frontend/coverage
+      - frontend/logs
+    reports:
+      junit: junit.xml
+    when: always
+  tags:
+    - hetzner-amd-beefy
+
+test:frontend:docs-links:
+  stage: pre-test
+  image: ${CI_DEPENDENCY_PROXY_DIRECT_GROUP_IMAGE_PREFIX}/${NODE_IMAGE}
+  rules:
+    - if: $CI_COMMIT_REF_NAME !~ /^(staging|saas-[a-zA-Z0-9.]+)$/
+    - changes:
+        paths: ['frontend/**/*']
+        compare_to: '${RULES_CHANGES_COMPARE_TO_REF}'
+  needs: []
+  before_script:
+    - apk add --no-cache curl
+    - cd frontend
+  script:
+    - links=$(grep -r docs.mender.io src/ | grep -v snapshots | sed -e 's,\${docsVersion},'${DOCS_VERSION}'/,g' | sed -e 's,\${path},''/,g' | egrep -o 'https?://[^ `"]+' | sort | uniq)
+    - error=0
+    - |
+      for url in $links; do
+        if ! curl --output /dev/null --silent --head --fail $url; then
+          error=1
+          echo "[ko] $url"
+        else
+          echo "[ok] $url"
+        fi
+      done
+      if [ $error -gt 0 ]; then
+        exit 1
+      fi
+  tags:
+    - hetzner-amd-beefy
+
+test:frontend:docs-links:hosted:
+  extends: test:frontend:docs-links
+  rules:
+    - if: $CI_COMMIT_REF_NAME =~ /^(staging|saas-[a-zA-Z0-9.]+)$/
+  variables:
+    DOCS_VERSION: hosted
+
+build:frontend:docker:
+  extends: .template:build:docker
+  rules:
+    - changes:
+        paths: ['frontend/**/*']
+        compare_to: '${RULES_CHANGES_COMPARE_TO_REF}'
+      when: always
+  script:
+    - cd frontend
+    - docker context create ci;
+    - docker builder create --name ci-builder ci;
+    - unset DOCKER_HOST;
+    - docker buildx create --name gui-builder --driver=docker-container ci
+    # needed for the license generation step
+    - docker build
+      --tag ${FRONTEND_REPOSITORY}:base
+      --target base
+      --builder=gui-builder
+      --output type=docker,dest=baseImage.tar
```

Could also be pushed to the internal registry to allow reuse of existing layers.
```suggestion
    - docker build
      --tag ${FRONTEND_REPOSITORY}:base
      --target base
      --builder=gui-builder
      --tag ${CI_REGISTRY_IMAGE}:frontend-license-${CI_PIPELINE_ID}
```